File: app/src/main/java/com/github/bjoernpetersen/q/ui/SuggestActivity.java
package com.github.bjoernpetersen.q.ui;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;
import com.ToxicBakery.viewpager.transforms.RotateUpTransformer;
import com.github.bjoernpetersen.jmusicbot.client.ApiException;
import com.github.bjoernpetersen.jmusicbot.client.model.NamedPlugin;
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry;
import com.github.bjoernpetersen.jmusicbot.client.model.Song;
import com.github.bjoernpetersen.q.QueueState;
import com.github.bjoernpetersen.q.R;
import com.github.bjoernpetersen.q.api.Auth;
import com.github.bjoernpetersen.q.api.AuthException;
import com.github.bjoernpetersen.q.api.Config;
import com.github.bjoernpetersen.q.api.Connection;
import com.github.bjoernpetersen.q.api.LoginException;
import com.github.bjoernpetersen.q.api.RegisterException;
import com.github.bjoernpetersen.q.ui.fragments.SongFragment;
import com.github.bjoernpetersen.q.ui.fragments.SuggestFragment;
import java.util.ArrayList;
import java.util.List;

public class SuggestActivity extends AppCompatActivity implements
    SuggestFragment.OnFragmentInteractionListener, SongFragment.OnListFragmentInteractionListener {

  private static final String TAG = SuggestActivity.class.getSimpleName();

  private ViewPager viewPager;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_search);
    setTitle(R.string.title_suggestions);

    ActionBar actionBar = getSupportActionBar();
    if (actionBar == null) {
      throw new IllegalStateException();
    }
    actionBar.setDisplayHomeAsUpEnabled(true);

    final ViewPager viewPager = (ViewPager) findViewById(R.id.view_pager);
    this.viewPager = viewPager;
    viewPager.setPageTransformer(true, new RotateUpTransformer());
    viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
      @Override
      public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
      }

      @Override
      public void onPageSelected(int position) {
        refreshSuggestions(position);
      }

      @Override
      public void onPageScrollStateChanged(int state) {
      }
    });

    loadSuggesters();
  }

  @Override
  protected void onDestroy() {
    viewPager = null;
    super.onDestroy();
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    MenuInflater inflater = getMenuInflater();
    inflater.inflate(R.menu.suggest_menu, menu);
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        finish();
        return true;
      case R.id.refresh:
        refreshSuggestions();
        return true;
      default:
        return super.onOptionsItemSelected(item);
    }
  }

  @Override
  protected void onStart() {
    super.onStart();
    if (!Config.INSTANCE.hasUser()) {
      startActivity(new Intent(this, LoginActivity.class));
    }
  }

  private void loadSuggesters() {
    new Thread(new Runnable() {
      @Override
      public void run() {
        try {
          final List<NamedPlugin> providers = Connection.INSTANCE.getSuggesters();
          runOnUiThread(new Runnable() {
            @Override
            public void run() {
              updateSuggesters(providers);
            }
          });
        } catch (ApiException e) {
          Log.e(TAG, "Could not retrieve providers", e);
          runOnUiThread(new Runnable() {
            @Override
            public void run() {
              Toast.makeText(
                  SuggestActivity.this,
                  getString(R.string.no_suggester_found),
                  Toast.LENGTH_SHORT
              ).show();
              finish();
            }
          });
        }
      }
    }, "SuggesterLoader").start();
  }

  private void updateSuggesters(List<NamedPlugin> suggesters) {
    if (viewPager != null) {
      viewPager
          .setAdapter(new SuggestFragmentPagerAdapter(getSupportFragmentManager(), suggesters));
    }
  }

  private void refreshSuggestions() {
    if (viewPager != null) {
      refreshSuggestions(viewPager.getCurrentItem());
    }
  }

  private void refreshSuggestions(int position) {
    PagerAdapter adapter = viewPager.getAdapter();
    if (adapter instanceof SuggestFragmentPagerAdapter) {
      SuggestFragmentPagerAdapter searchAdapter = (SuggestFragmentPagerAdapter) adapter;
      searchAdapter.getItem(position).update();
    }
  }

  @Override
  public void onAdd(final Song song) {
    new Thread(new Runnable() {
      @Override
      public void run() {
        try {
          String token = Auth.INSTANCE.getApiKey().getRaw();
          List<QueueEntry> queueEntries = Connection.INSTANCE
              .enqueue(token, song.getId(), song.getProvider().getId());
          QueueState.getInstance().set(queueEntries);
        } catch (ApiException e) {
          if (e.getCode() == 401) {
            Auth.INSTANCE.clear();
            Log.v(TAG, "Could not add song, trying again with cleared auth...");
            this.run();
          } else {
            Log.d(TAG, "Could not add song: " + e.getCode(), e);
          }
        } catch (RegisterException e) {
          if (e.getReason() == RegisterException.Reason.TAKEN) {
            runOnUiThread(new Runnable() {
              @Override
              public void run() {
                startActivity(new Intent(SuggestActivity.this, LoginActivity.class));
              }
            });
          }
        } catch (LoginException e) {
          runOnUiThread(new Runnable() {
            @Override
            public void run() {
              startActivity(new Intent(SuggestActivity.this, LoginActivity.class));
            }
          });
        } catch (AuthException e) {
          Log.d(TAG, "Could not add song", e);
        }
      }
    }, "enqueueThread").start();
  }

  @Override
  public boolean showAdd(Song song) {
    return true;
  }
}

final class SuggestFragmentPagerAdapter extends FragmentPagerAdapter {

  private final List<SuggestFragment> fragments;

  SuggestFragmentPagerAdapter(FragmentManager fm, List<NamedPlugin> suggesters) {
    super(fm);
    fragments = new ArrayList<>(suggesters.size());
    for (NamedPlugin provider : suggesters) {
      fragments.add(SuggestFragment.newInstance(provider));
    }
  }

  /**
   * Return the Fragment associated with a specified position.
   */
  @Override
  public SuggestFragment getItem(int position) {
    return fragments.get(position);
  }

  /**
   * Return the number of views available.
   */
  @Override
  public int getCount() {
    return fragments.size();
  }

  @Override
  public CharSequence getPageTitle(int position) {
    return getItem(position).getSuggester().getId();
  }
}

File: app/src/main/java/com/github/bjoernpetersen/q/ui/SuggestActivity.kt
package com.github.bjoernpetersen.q.ui

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import com.ToxicBakery.viewpager.transforms.RotateUpTransformer
import com.github.bjoernpetersen.jmusicbot.client.ApiException
import com.github.bjoernpetersen.jmusicbot.client.model.NamedPlugin
import com.github.bjoernpetersen.jmusicbot.client.model.Song
import com.github.bjoernpetersen.q.QueueState
import com.github.bjoernpetersen.q.R
import com.github.bjoernpetersen.q.api.*
import com.github.bjoernpetersen.q.tag
import com.github.bjoernpetersen.q.ui.fragments.CachedFragmentPagerAdapter
import com.github.bjoernpetersen.q.ui.fragments.SongFragment
import com.github.bjoernpetersen.q.ui.fragments.SuggestFragment
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_search.*
import java.lang.ref.WeakReference

class SuggestActivity : AppCompatActivity(), SuggestFragment.OnFragmentInteractionListener,
    SongFragment.SongFragmentInteractionListener, ObserverUser {

  override lateinit var observers: MutableList<WeakReference<Disposable>>
  private var suggesters: List<NamedPlugin> = emptyList()
    set(value) {
      if (field != value) {
        field = value
        view_pager?.adapter = SuggestFragmentPagerAdapter(supportFragmentManager, value)
        supportFragmentManager.executePendingTransactions()
        refreshSuggestions()
      }
    }
  private val activeSuggester: NamedPlugin?
    get() {
      val index = view_pager.currentItem
      return if (index < suggesters.size) suggesters[index]
      else null
    }

  override fun initObservers() {
    observers = ArrayList()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_search)
    setTitle(R.string.title_suggestions)

    val actionBar = supportActionBar ?: throw IllegalStateException()
    actionBar.setDisplayHomeAsUpEnabled(true)

    view_pager.setPageTransformer(true, RotateUpTransformer())
    view_pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
      override fun onPageScrolled(position: Int, positionOffset: Float,
          positionOffsetPixels: Int) {
      }

      override fun onPageSelected(position: Int) = refreshSuggestions()
      override fun onPageScrollStateChanged(state: Int) {}
    })
  }

  override fun onDestroy() {
    view_pager.clearOnPageChangeListeners()
    super.onDestroy()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean = true.also {
    menuInflater.inflate(R.menu.suggest_menu, menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
    android.R.id.home -> {
      finish()
      true
    }
    R.id.refresh -> {
      refreshSuggestions()
      true
    }
    else -> super.onOptionsItemSelected(item)
  }

  override fun onStart() {
    super.onStart()
    if (!Config.hasUser()) {
      startActivity(Intent(this, LoginActivity::class.java))
      return
    }
    initObservers()
    loadSuggesters()
  }

  override fun onStop() {
    disposeObservers()
    super.onStop()
  }

  override fun onResume() {
    super.onResume()
    checkWifiState()
  }

  private fun loadSuggesters() {
    Observable.fromCallable { Connection.getSuggesters() }
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          suggesters = it
        }, {
          Log.d(tag(), "Could not retrieve suggesters", it)
          Toast.makeText(this, getString(R.string.no_suggester_found), Toast.LENGTH_SHORT).show()
          finish()
        }).store()
  }

  private fun refreshSuggestions() {
    (view_pager.adapter as? SuggestFragmentPagerAdapter)
        ?.getFragment(view_pager.currentItem)
        ?.refresh()
  }

  override fun onContextMenu(song: Song, menuItem: MenuItem, enable: (Boolean) -> Unit): Boolean {
    when (menuItem.itemId) {
      R.id.enqueue_button -> onClick(song, enable)
      R.id.remove_button -> dislike(song, enable)
    }
    return true
  }

  private fun dislike(song: Song, enable: (Boolean) -> Unit) {
    enable(false)
    val suggesterId = activeSuggester?.id ?: return
    Observable.fromCallable { Auth.apiKey.raw }
        .subscribeOn(Schedulers.io())
        .map { Connection.removeSuggestion(suggesterId, it, song.id, song.provider.id) }
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          Log.d(tag(), "Successfully disliked song.")
        }, {
          enable(true)
          Toast.makeText(this, R.string.dislike_error, Toast.LENGTH_SHORT).show()
          Log.d(tag(), "Could not remove suggestion", it)
        })
        .store()
  }

  override fun onClick(song: Song, enable: (Boolean) -> Unit) {
    enable(false)
    Observable.fromCallable { Auth.apiKey.raw }
        .subscribeOn(Schedulers.io())
        .map { Connection.enqueue(it, song.id, song.provider.id) }
        .retry(1, {
          if (it is ApiException && it.code == 401) {
            Auth.clear(); true
          } else false
        })
        .doOnNext { QueueState.queue = it }
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          Log.d(tag(), "Successfully added song to queue: ${song.title}")
        }, {
          Log.d(tag(), "Could not add a song.")
          enable(true)
          when (it) {
            is RegisterException -> if (it.reason == RegisterException.Reason.TAKEN) {
              Toast.makeText(
                  this,
                  "Your username is already taken.",
                  Toast.LENGTH_SHORT
              ).show()
              startActivity(Intent(this, LoginActivity::class.java))
            }
            is LoginException -> if (it.reason == LoginException.Reason.WRONG_UUID
                || it.reason == LoginException.Reason.WRONG_PASSWORD
                || it.reason == LoginException.Reason.NEEDS_AUTH) {
              Toast.makeText(
                  this,
                  "Can't login with current username and password.",
                  Toast.LENGTH_SHORT
              ).show()
              startActivity(Intent(this, LoginActivity::class.java))
            }
          }
        })
        .store()
  }

  override fun isEnabled(song: Song): Boolean = !QueueState.queue.map { it.song }.any { it == song }

  override fun isEnabled(song: Song, menuItemId: Int): Boolean = when (menuItemId) {
    R.id.enqueue_button -> isEnabled(song)
    R.id.remove_button -> Auth.hasPermissionNoRefresh(Permission.DISLIKE)
    else -> true
  }
}

internal class SuggestFragmentPagerAdapter(fm: FragmentManager,
    private val suggesters: List<NamedPlugin>) :
    CachedFragmentPagerAdapter<SuggestFragment>(fm, suggesters.size) {

  /**
   * Return the Fragment associated with a specified position.
   */
  override fun getItem(position: Int): SuggestFragment =
      SuggestFragment.newInstance(suggesters[position])

  /**
   * Return the number of views available.
   */
  override fun getCount(): Int = suggesters.size

  override fun getPageTitle(position: Int): CharSequence = suggesters[position].name
}